### PR TITLE
Java 9+ support

### LINKED ISSUE #1943: Problem with Java destinations JDK9 and above
# syslog-ng

## Version of syslog-ng

syslog-ng 3 (3.14.1)
Config version: 3.14
Installer-Version: 3.14.1
Revision: 
Compile-Date: Feb 28 2018 07:34:20
Module-Directory: //usr/lib64/syslog-ng
Module-Path: //usr/lib64/syslog-ng
Available-Modules: afsocket,json-plugin,cef,stardate,graphite,disk-buffer,sdjournal,afuser,afamqp,afprog,date,linux-kmsg-format,system-source,map-value-pairs,tags-parser,affile,cryptofuncs,xml,dbparser,appmodel,kvformat,snmptrapd-parser,basicfuncs,add-contextual-data,confgen,mod-java,afstomp,syslogformat,pseudofile,csvparser,tfgetent
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: on
Enable-TCP-Wrapper: on
Enable-Linux-Caps: on
Enable-Systemd: on

## Platform

Linux statia 3.10.0-693.5.2.el7.x86_64 #1 SMP Fri Oct 20 20:32:50 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
Centos 7.4

# Issue

I tried to start elasticsearch2 with jre 9.0.4 , seems the elastic2 destination is not jdk9 ready yet.
Same configuration works perfect with jre1.8.0_161-b12
RPMs are from https://copr-be.cloud.fedoraproject.org/results/czanik/syslog-ng314/

## Failure

Fails to start with java exception on stdout:

WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.syslog_ng.SyslogNgClassLoader (file:/usr/lib64/syslog-ng/java-modules/syslog-ng-core.jar) to method java.net.URLClassLoader.addURL(java.net.URL)
WARNING: Please consider reporting this to the maintainers of org.syslog_ng.SyslogNgClassLoader
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[2018-03-27T17:01:53.585521] Error while expanding path list: object is not an instance of declaring class;
java.lang.IllegalArgumentException: object is not an instance of declaring class
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.syslog_ng.SyslogNgClassLoader.expandClassPath(SyslogNgClassLoader.java:144)
	at org.syslog_ng.SyslogNgClassLoader.loadClass(SyslogNgClassLoader.java:65)
[2018-03-27T17:01:53.586889] Can't find class; class_name='org.syslog_ng.elasticsearch_v2.ElasticSearchDestination'
[2018-03-27T17:01:53.589907] Error initializing message pipeline; plugin name='java', location='#buffer:2:3'

## Steps to reproduce

1. LD_LIBRARY_PATH=<JRE9_DIR>/lib/server:${LD_LIBRARY_PATH}
2. export LD_LIBRARY_PATH
3. start syslog-ng

## Configuration
<stipped down snippet>

source s_global_raw {
	channel {
		source {
			systemd-journal(
				log-fetch-limit(1000)
				default-level("info")
			);
			internal();
			udp(ip(127.0.0.1) port(514));
		};
	};
};

destination d_elastic_local {
	elasticsearch2(
		index("logs-${YEAR}.${MONTH}.${DAY}")
		type("my-log")
		cluster("elasticsearch")
		persist_name("elastic-local")
		template("$(format-json --scope nv-pairs @timestamp=${S_ISODATE})")
		client-lib-dir("/usr/share/elasticsearch/lib")
		flush-limit(2000)
		client-mode("http")
		cluster-url("http://localhost:9200")
	);
};

log {
	source(s_global_raw);
	destination(d_elastic_local);
};

### LINKED ISSUE #3053: compile issue with java 11
make
```
  GEN      modules/java/org_syslog_ng_LogMessage.h
/bin/bash: /root/syslog-ng/: Is a directory
make: *** [Makefile:26085: modules/java/org_syslog_ng_LogMessage.h] Error 126
```

### PR DESCRIPTION
**Pull request by MrAnno**
- `javah` has been removed from JDK 10, `javac -h` should be used instead
- fix illegal reflection issue with Java 9+:

The system class loader of syslog-ng has been replaced with our own implementation, which extends the default URLClassLoader.

That way, the protected addURL() method can be used directly without reflection.

Please note that URLClassLoader::addURL() is protected for a good reason.
Once a URL-based class loader is initialized, we should not really modify its class paths. This is still being done in this PR, a real fix would restructure our Java module's initialization logic so that we don't have to extend our class paths more than once (we'd probably have to fix Java VM reinitialization during reload first).

Fixes #3053
Fixes #1943

----

Requires testing:
- [x] internal regression tests

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by MrAnno** [comment]
Requires testing:
- internal regression tests
- manual test with Java 18 using a real module (HDFS, or retest https://github.com/syslog-ng/syslog-ng/pull/3432#issuecomment-702617259)

File: modules/java/src/main/java/org/syslog_ng/SyslogNgClassLoader.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ public Class loadClass(String className, String pathList) {
       if (urls.length > 0) {
         expandClassPath(urls);
       }
-      result = Class.forName(className, true, classLoader);
+      result = Class.forName(className, true, this);
```
Comment (MrAnno) [inline modules/java/src/main/java/org/syslog_ng/SyslogNgClassLoader.java:67]: `this` -> `Thread.currentThread().getContextClassLoader()`

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

File: modules/java/native/java_machine.c [hunk 2]
```diff
@@ -191,6 +191,9 @@ _setup_jvm_options_array(JavaVMSingleton *self, const gchar *jvm_options_str)
   jvm_options_array = _jvm_options_array_append(jvm_options_array,
                                                 g_strdup_printf("-Djava.system.class.loader=org.syslog_ng.SyslogNgClassLoader"));
 
+  jvm_options_array = _jvm_options_array_append(jvm_options_array,
```
Comment (Kokan) [inline modules/java/native/java_machine.c:194]: This should be added to the `predefined_options` as well.

File: packaging/debian/control [hunk 3]
```diff
@@ -37,7 +37,7 @@ Build-Depends: debhelper (>= 10~),
                libsnappy-dev,
                libsnmp-dev [linux-any hurd-i386],
                librdkafka-dev (>= 1.1.0) <!sng-nokafka>,
-               openjdk-8-jdk <!nojava>,
+               openjdk-11-jdk <!nojava>,
```
Comment (Kokan) [inline packaging/debian/control:40]: Could we use `default-jdk-headless` instead ? The goal is not to specify java version, and headless just why not.

Comment (MrAnno) [reply packaging/debian/control:40]: The lighter, the better. :)

File: modules/java/native/java_machine.c (hunk 2 above)
Comment (MrAnno) [reply modules/java/native/java_machine.c:194]: Fixed, thank you.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Review by Kokan — APPROVED** [review]

**Comment by MrAnno** [comment]
I've rebased this PR against the current master. We've previously run internal tests on this branch, and we had 2 reviewers before the rebase.

There is still a refcache issue with newer Java versions, but I think the PR could go in without that fix because it improves the current situation a lot anyway.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
@kira-syslogng retest this please

**Comment by bazsi** [comment]
> I've rebased this PR against the current master. We've previously run internal tests on this branch, and we had 2 reviewers before the rebase.
> 
> There is still a refcache issue with newer Java versions, but I think the PR could go in without that fix because it improves the current situation a bit anyway.

I am approving this based on the two earlier reviewers and that it does not touch anything material outside of modules/java and java-modules.

**Review by bazsi — APPROVED** [review]

**Comment by bazsi** [comment]
I've updated NEWS.md with this sentence:

+  * `java()` destinations: fixed compatibility with newer Java versions,
+    syslog-ng is now able to compile up to Java 18.
+


If you have any more details to add, please let me know.

**Pull request merged**